### ISSUE TITLE
Predicting binary outcomes with R

### ISSUE THREAD

**Issue by juliettetroadec** [juliettetroadec/hello-word#2]
**Contents**: gbm (Generalized Boosted Models), glmnet (General Linear Models)
**Libraries**: {caret}, {pROC}
**Data**: Titanic dataset 

**1. Data cleaning**
```
df <- read.csv('http://math.ucdenver.edu/RTutorial/titanic.txt',sep='\t')
print(str(df))
```
```
## 'data.frame':	1313 obs. of  5 variables:
##  $ Name    : Factor w/ 1310 levels "Abbing, Mr Anthony",..: 22 25 26 27 24 31 45 46 50 54 ...
##  $ PClass  : Factor w/ 3 levels "1st","2nd","3rd": 1 1 1 1 1 1 1 1 1 1 ...
##  $ Age     : num  29 2 30 25 0.92 47 63 39 58 71 ...
##  $ Sex     : Factor w/ 2 levels "female","male": 1 1 2 1 2 2 1 2 1 2 ...
##  $ Survived: int  1 0 0 0 1 1 1 0 1 0 ...
## NULL
```
The Name variable contains factor titles that we want to extract (dropping names)

```
df$Title <- ifelse(grepl('Mr ',df$Name),'Mr',ifelse(grepl('Mrs ',df$Name),'Mrs',ifelse(grepl('Miss',df$Name),'Miss','Nothing')))
```

The Age variable has missing values that we replace by median age (some analysis functions do not handle missing values, like **glmnet**)

```
data$Age[is.na(data$Age)] <- median(data$Age, na.rm=T)
```
We want the target variable to be the last column of the dataset.

```
df <- df[c('PClass', 'Age',    'Sex',   'Title', 'Survived')]
print(str(df))
```
```
## 'data.frame':	1313 obs. of  5 variables:
##  $ PClass  : Factor w/ 3 levels "1st","2nd","3rd": 1 1 1 1 1 1 1 1 1 1 ...
##  $ Age     : num  29 2 30 25 0.92 47 63 39 58 71 ...
##  $ Sex     : Factor w/ 2 levels "female","male": 1 1 2 1 2 2 1 2 1 2 ...
##  $ Title   : chr  "Miss" "Miss" "Mr" "Mrs" ...
##  $ Survived: int  1 0 0 0 1 1 1 0 1 0 ...
## NULL
```

**2. Factor variables**

**gbm** can deal with factor variables since it dummify them internaly, but **glmnet** can't, it only accepts numeric data. Let's dummify factors and harmonize everything. Here we use the {caret} package and particularly the function **dummyVars**:
```
df$Title <- as.factor(df$Title)
dfDummy <- dummyVars("~.",data=df, fullRank=F)
dfDF <- as.data.frame(predict(titanicDummy,df))
print(names(df))
```
Each unique factor is separated into its own value.

**3. Outcome variable**
```
prop.table(table(df$Survivied))
```
```
## 
##      0      1 
## 0.6573 0.3427
```
34.27% of the dataset contains survivors of the Titanic tragedy. If the proportion of the studied event is less than 15%, then we say that it is a rare event.
Let's generalize variables:
```
outcomeName <- 'Survived'
predictorsNames <- names(df)[names(df) != outcomeName]
```

**4. Modeling**

The {caret} package supports a lot of models:
```
names(getModelInfo())
```

**a. Gbm model**

What types of modeling does the {caret} package function **gbm** support?
```
getModelInfo()$gbm$type
```
```
## [1] "Regression"     "Classification"
```
Here we want to predict a binary variable, so we will need the classificaiton mode. We do this by changing the target variable to a factor (we use a copy of the outcome as we’ll need the original one for our next model):
```
df$Survived2 <- ifelse(df$Survived==1,'yes','no')
df$Survived2 <- as.factor(df$Survived2)
outcomeName <- 'Survived2'
```

Now we want to split the data into a train set and a test set
```
set.seed(94)
ind <- sample(nrow(df), 0.75*nrow(df), replace = FALSE)
train <- df[ind,]
test <- df[-ind,]
```
Another option is to use the createDataPartition function the {caret} package offers.
```
set.seed(94)
ind <- createDataPartition(df[,outcomeName], p = .75, list = FALSE, times = 1)
train <- df[ ind,]
test  <- df[-ind,]
```
{caret} offers many tuning functions: the **trainControl** function allows to control the data resampling by spliting the training data set internally into train/test sets to figure out the best model settings.  In this case, we’re going to **cross-validate** the data 3 times - training it 3 times on different data portions before settling on the best tuning parameters (for **gbm** it is trees, shrinkage, and interaction depth). 
```
control <- trainControl(method='cv', number=3, returnResamp='none', summaryFunction = twoClassSummary, classProbs = TRUE)
```
Time to teach the model how to recognize Titanic survivirs. Because we need a classification rather than a regression, we set **'ROC'** metric instead of the default 'RMSE' in the function:
```
gbm.model <- train(train[,predictorsNames], train[,outcomeName], method='gbm', trControl=control, metric = "ROC", preProc = c("center", "scale"))
```
```
## Iter   TrainDeviance   ValidDeviance   StepSize   Improve
##      1        1.2314             nan     0.1000    0.0245
##      2        1.1948             nan     0.1000    0.0192
##      3        1.1594             nan     0.1000    0.0158
...
```
To find out what variables are the most important, we call the summary() function:
```
summary(gbm.model)
```
![capture d ecran 2016-11-12 a 11 12 01](https://cloud.githubusercontent.com/assets/17045683/20237091/ee03b838-a8c8-11e6-8085-7554d4a61acb.png)
```
##                         var rel.inf
## Title.Mr           Title.Mr 26.2756
## PClass.3rd       PClass.3rd 20.8523
## Sex.male           Sex.male 20.7569
## Sex.female       Sex.female 11.4357
## Age                     Age 10.3042
## PClass.1st       PClass.1st  8.2905
## Title.Mrs         Title.Mrs  1.7515
## Title.Miss       Title.Miss  0.3332
## PClass.2nd       PClass.2nd  0.0000
## Title.Nothing Title.Nothing  0.0000
```

What are the best tuning parameters for this model?
```
print(model)
```
```
## Stochastic Gradient Boosting 
## 
## 986 samples
##  10 predictor
##   2 classes: 'nope', 'yes' 
## 
## Pre-processing: centered, scaled 
## Resampling: Cross-Validated (3 fold) 
...
## Tuning parameter 'shrinkage' was held constant at a value of 0.1
## ROC was used to select the optimal model using  the largest value.
## The final values used for the model were n.trees = 100,
##  interaction.depth = 1 and shrinkage = 0.1.
```
Here we need **100 trees**, **interaction.depth = 1** and **shrinkage = 0.1**.

**b. Evaluate gbm model**

We can evaluate the model using two methods: 
- **raw**: gives a class prediction (here 'yes' or 'no')
- **prob**: gives the probability on how sure the model is about the event
Prob is a good solution since we can set the threshold (event='yes' if prob>0.5 for instance) and study AUC score (which requires probabilities, not classes). There are situations in which having class values is valuable, such as when using multinomial models where you’re predicting more than two values.

It's time to use the **predict** function. 
```
class.prediction <- predict(object=model, test[,predictorsNames], type='raw')
head(class.prediction)
```
```
## [1] yes  no yes  no no no
## Levels: no yes
```
Let’s start by looking at class predictions and use the caret **postResample** function to get an accuracy score:
```
print(postResample(pred=prediction, obs=as.factor(test[,outcomeName])))
```
```
## Accuracy    Kappa 
##   0.8135   0.5644
```
The model is correct 81.35% of the time, which is good.
Looking now at probabilities:
```
# probabilites 
library(pROC)
prob.prediction <- predict(object=model, test[,predictorsNames], type='prob')
head(prob.prediction)
```
```
##      nope    yes
## 1 0.07292 0.9271
## 2 0.76058 0.2394
## 3 0.43309 0.5669
## 4 0.67279 0.3272
## 5 0.67279 0.3272
```
To get the AUC score, we need to pass the *yes column* to the roc function (each row adds up to 1 but we’re interested in the yes, the survivors):
```
auc <- roc(ifelse(test[,outcomeName]=="yes",1,0), predictions[[2]])
print(auc$auc)
```
```
## Area under the curve: 0.825
```
The model has a 0.825 AUC score.
**Hint:** an AUC ranges between 0.5 and 1, where 0.5 is random and 1 is perfect.

**c. Glmnet Model**

Let's try a regression model. What modeling support does **glmnet** function offer?
```
getModelInfo()$glmnet$type
```
```
## [1] "Regression"     "Classification"
```
Spliting the dataset into train and test sets:
```
outcomeName <- 'Survived'
set.seed(94)
ind <- sample(nrow(df), 0.75*nrow(df), replace=FALSE)
traindf <- df[ind,]
testdf  <- df[-ind,]
```
Re-running some of the basic training and prediction functions with some slight changes:
```
control <- trainControl(method='cv', number=3, returnResamp='none')
glmnet.model <- train(traindf[,predictorsNames], traindf[,outcomeName], method='glmnet',  metric = "RMSE", trControl=control)
```
Note that we use 'RMSE' option instead of 'ROC'.
```
prediction <- predict(object=glmnet.model, testdf[,predictorsNames])
auc <- roc(testdf[,outcomeName], prediction)
print(auc$auc)
```
```
## Area under the curve: 0.857
```
The AUC score is higher than in the previous gbm model. 

The **varlmp** {caret} function helps figuring out what variables are important in model, this is a great feature of the glmnet model. It returns positive and negative variable importances unlike most models. This allows to deepen variable understanding (e.g. being in PClass.1st increases the probability of survival while being in PClass.3rd decreases the event probability).
**Note: make sure to set scale to False**

```
plot(varImp(glmnet.model, scale=FALSE))
```
![capture d ecran 2016-11-12 a 11 51 31](https://cloud.githubusercontent.com/assets/17045683/20237300/633d63a6-a8ce-11e6-8274-145487cd6124.png)

**Comment by juliettetroadec** [comment]
Good tutorial to begin with classification.
Source: http://amunategui.github.io/binary-outcome-modeling/
